atcoder/internal_math.hpp

```diff
@@ -177,6 +177,35 @@ constexpr int primitive_root_constexpr(int m) {
 }
 template <int m> constexpr int primitive_root = primitive_root_constexpr(m);
 
+// @param n `n < 2^32`
+// @param m `1 <= m < 2^32`
+// @return sum_{i=0}^{n-1} floor((ai + b) / m) (mod 2^64)
+unsigned long long floor_sum_unsigned(unsigned long long n,
+                                      unsigned long long m,
+                                      unsigned long long a,
+                                      unsigned long long b) {
+    unsigned long long ans = 0;
+    while (true) {
+        if (a >= m) {
+            ans += n * (n - 1) / 2 * (a / m);
+            a %= m;
+        }
+        if (b >= m) {
+            ans += n * (b / m);
+            b %= m;
+        }
+
+        unsigned long long y_max = a * n + b;
```

Comment:
If `y_max = a * n + b >= 2^64`, are `y_max / m` and `y_max % m` calculated properly?

Reply:
a * n + b <= (m - 1) * n + (m - 1) <= (m - 1) * (n + 1) <= (2 ^ 32 - 2) * 2 ^ 32 < 2 ^ 64

Reply:
Aha! I forgot the constraints about `m`, sorry.